src/devices/sound/ym2203.cpp

```diff
@@ -0,0 +1,233 @@
+// license:BSD-3-Clause
+// copyright-holders:Aaron Giles
+
+#include "emu.h"
+#include "ym2203.h"
+
+
+DEFINE_DEVICE_TYPE(YM2203, ym2203_device, "ym2203", "YM2203 OPN")
+
+
+//*********************************************************
+//  INLINE HELPERS
+//*********************************************************
+
+//-------------------------------------------------
+//  linear_to_fp - given a 32-bit signed input
+//  value, convert it to a signed 10.3 floating-
+//  point value
+//-------------------------------------------------
+
+inline s16 linear_to_fp(s32 value)
+{
+	// start with the absolute value
+	s32 avalue = (value < 0) ? -value : value;
+	u8 shift = 0;
+
+	// shift until absolute value fits in 9 bits (bit 10 is the sign)
+	while (avalue != (avalue & 0x1ff))
+	{
+		avalue >>= 1;
+		shift++;
+	}
+
+	// if out of range, just return maximum; note that YM3012 DAC does
+	// not support a shift count of 7, so we clamp at 6
+	if (shift >= 7)
+		shift = 6, avalue = 0x1ff;
+
+	// encode with shift in low 3 bits and signed mantissa in upper
+	return shift | (((value < 0) ? -avalue : avalue) << 3);
+}
+
+
+//-------------------------------------------------
+//  fp_to_linear - given a 10.3 floating-point
+//  value, convert it to a signed 16-bit value,
+//  clamping
+//-------------------------------------------------
+
+inline s32 fp_to_linear(s16 value)
+{
+	return (value >> 3) << BIT(value, 0, 3);
+}
+
+
+
+//*********************************************************
+//  YM2203 DEVICE
+//*********************************************************
+
+//-------------------------------------------------
+//  ym2203_device - constructor
+//-------------------------------------------------
+
+ym2203_device::ym2203_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	ay8910_device(mconfig, YM2203, tag, owner, clock, PSG_TYPE_YM, 3, 2),
+	m_opn(*this),
+	m_stream(nullptr),
+	m_address(0)
+{
+}
+
+
+//-------------------------------------------------
+//  read - handle a read from the device
+//-------------------------------------------------
+
+u8 ym2203_device::read(offs_t offset)
+{
+	u8 result = 0;
+	switch (offset & 1)
+	{
+		case 0:	// status port
+			result = m_opn.status();
+			break;
+
+		case 1: // data port (only SSG)
+			if (m_address < 0x10)
+				result = ay8910_read_ym();
+			break;
+	}
+	return result;
+}
+
+
+//-------------------------------------------------
+//  write - handle a write to the register
+//  interface
+//-------------------------------------------------
+
+void ym2203_device::write(offs_t offset, u8 value)
+{
+	switch (offset & 1)
+	{
+		case 0:	// address port
+			m_address = value;
+
+			// write register to SSG
+			if (m_address < 0x10)
+				ay8910_write_ym(0, m_address);
+
+			// prescaler select : 2d,2e,2f
+			else if (m_address == 0x2d)
```

Comment:
Hmmm, trying to think how to work around this. In most cases where I'm doing it, it's a series of special cases, and I'd rather have an explanation for each special case next to its conditional, rather than in a block at the top. If I put braces around the single-line ifs would that be sufficient?

Reply:
I can read the source code fine

Reply:
Strictly speaking, the "else" in your code here is not necessary, and neither is my example https://github.com/mamedev/mame/blob/master/src/devices/cpu/cops1/cops1d.cpp#L139

blank line or not, the code is readable.

Reply:
After inheriting a bunch of code like this in MAME’s core, I’ve come to respect coding standards that say things like “always use braces when there are multiple lines between the `if` and `else`”.

I’m fine with any of these styles in Allman-style code:
```
if (cond1) // comment for case 1
    statement1;
else if (cond2) // comment for case 2
    statement2;

if (cond1)
    // comment for case 1
    statement1;
else if (cond2)
    // comment for case 2
    statement2;

if (cond1)
{ // comment for case 1
    statement1;

}
else if (cond2)
{ // comment for case 2
    statement2;

}
```

I’m fine with this K&R style formats:
```
if (cond1) { // comment for case 1
    statement1;

} else if (cond2) { // comment for case 2
    statement2;
}
```

The ones I have a problem with are the ones where there’s a visual break before an `else` without braces around the controlled block, or a visual break between the closing brace and the else, i.e. these ones:
```
// comment for case 1
if (cond1)
    statement1;

// comment for case 2
else if (cond2)
    statement2;

// comment for case 1
if (cond1)
{
    statement1;
}
// comment for case 2
else if (cond2)
{
    statement2;
}
```

Reply:
Ok thanks. Don't really agree this improves much, but I've gone ahead and moved the comments inside braced blocks.